Update RMA functional tests

avinashkethineedi opened the pull request:
Added multi-workgroup support and updated RMA metric calculations

- Added multi-workgroup support to primitive tests (thread, wavefront, and workgroup)
- Updated RMA functional tests to capture timing after the `quiet` call

avinashkethineedi committed:
Update primitive tests for multi-workgroup support

avinashkethineedi committed:
Update workgroup primitive tests for multi-workgroup support

avinashkethineedi committed:
Update workfront primitive tests for multi-workgroup support

avinashkethineedi committed:
Update team based primitive tests for multi-workgroup support

avinashkethineedi committed:
Update RMA functional tests to capture timing after quiet call

- Modified RMA functional tests to record the time after a `quiet` call in thread, wavefront, and workgroup RMA calls.

Yiltan commented on tests/functional_tests/swarm_tester.cpp:
```diff
@@ -49,7 +49,7 @@ __global__ void GetSwarmTest(int loop, int skip, long long int *start_time,
     if (i == skip) {
       start_time[wg_id] = wall_clock64();
```
No `rocshmem_ctx_quiet()`? It was added to other tests but not here, is it not needed for this test?

Yiltan commented on tests/functional_tests/wavefront_primitives.cpp:
```diff
@@ -31,54 +31,56 @@ using namespace rocshmem;
-  if (hipThreadIdx_x == 0) {
-    end_time[hipBlockIdx_x] = wall_clock64();
+  if(is_thread_zero_in_wave()) {
+    rocshmem_ctx_quiet(ctx);
```
Minor comment: `if(` -> `if (`

Do we need a `__syncthreads()`. E.g. if wave 0 arrives first and calls quiet before wave n has made any calls. Or does quiet handle this issue. Some comment applies to the other new additions of quiet

avinashkethineedi replied:
`__syncthreads` is not required for wavefront primitives; thread zero in each wavefront will call `quiet` and capture the end time, independent of the other wavefronts in the workgroup.

edgargabriel commented on tests/functional_tests/team_ctx_primitive_tester.cpp:
```diff
@@ -80,18 +87,29 @@ __global__ void TeamCtxPrimitiveTest(int loop, int skip, long long int *start_ti
+  dest = (char *)rocshmem_malloc(buff_size);
+
+  if (source == nullptr || dest == nullptr) {
+    std::cout << "Error allocating memory from symmetric heap" << std::endl;
```
shouldn't error messages use `std::cerr` instead of `std::cout` ?

edgargabriel commented on tests/functional_tests/wavefront_primitives.cpp:
```diff
@@ -88,48 +90,58 @@ __global__ void WaveLevelPrimitiveTest(int loop, int skip,
+
+  if (source == nullptr || dest == nullptr) {
+    std::cout << "Error allocating memory from symmetric heap" << std::endl;
+    std::cout << "source: " << source << ", dest: " << dest << std::endl;
```
same here, we should probably use `std::cerr`

edgargabriel commented on tests/functional_tests/workgroup_primitives.cpp:
```diff
@@ -86,45 +86,57 @@ __global__ void ExtendedPrimitiveTest(int loop, int skip,
+
+  if (source == nullptr || dest == nullptr) {
+    std::cout << "Error allocating memory from symmetric heap" << std::endl;
+    std::cout << "source: " << source << ", dest: " << dest << std::endl;
```
and one more

BKP reviewed (dismissed):
I'm not sure that swarm test is needed anymore.

avinashkethineedi committed:
Improve error handling and memory management

- Replaced `cout` with `cerr` for improved error reporting.
- Ensured all allocated memory is freed when `rocshmem_malloc` fails.

avinashkethineedi committed:
Update start time in primitive tests and latency calculations

- Modified primitive tests to capture the earliest start time.
- Updated latency calculations in functional tests.

avinashkethineedi committed:
Remove `GetSwarmTester`

avinashkethineedi committed:
Update start time in team primitive tests

BKP commented on tests/functional_tests/team_ctx_primitive_tester.cpp:
```diff
@@ -76,8 +86,24 @@ __global__ void TeamCtxPrimitiveTest(int loop, int skip, long long int *start_ti
+   */
   end_time[wg_id] = wall_clock64();
 
+  // Find the earliest start time
```
__syncthreads();

BKP commented on tests/functional_tests/team_ctx_primitive_tester.cpp:
```diff
@@ -76,8 +86,24 @@ __global__ void TeamCtxPrimitiveTest(int loop, int skip, long long int *start_ti
+      wf_start_time[t_id] = min(wf_start_time[t_id] , wf_start_time[t_id + i]);
+    }
+  }
+
```
__syncthreads();

BKP commented on tests/functional_tests/team_ctx_primitive_tester.cpp:
```diff
@@ -76,8 +86,24 @@ __global__ void TeamCtxPrimitiveTest(int loop, int skip, long long int *start_ti
 
+  // Find the earliest start time
+  int num_wfs = (get_flat_block_size() - 1 ) / wf_size + 1;
+  for (int i = num_wfs; i > 0; i >>= 1 ) {
```
num_wfs/2
Need the wf_start_time[t_id + i] to not go out of bounds. // bounds 16 hardcoded above

BKP commented on tests/functional_tests/wavefront_primitives.cpp:
```diff
@@ -31,54 +31,56 @@ using namespace rocshmem;
+    rocshmem_ctx_quiet(ctx);
+    end_time[idx] = wall_clock64();
   }
 
```
__syncthreads();

avinashkethineedi committed:
Invoke quiet call from a single thread within a block on a rocshmem context
